js/data/buffer2.js

```diff
@@ -0,0 +1,341 @@
+'use strict';
+
+// Note: all "sizes" are measured in bytes
+
+var util = require('../util/util');
+
+/**
+ * The `Buffer` class is responsible for managing one instance of `ArrayBuffer`. `ArrayBuffer`s
+ * provide low-level read/write access to a chunk of memory. `ArrayBuffer`s are populated with
+ * per-vertex data, uploaded to the GPU, and used in rendering.
+ *
+ * `Buffer` provides an abstraction over `ArrayBuffer`, making it behave like an array of
+ * statically typed structs. A buffer is comprised of items. An item is comprised of a set of
+ * attributes. Attributes are defined when the class is constructed.
+ *
+ * @class Buffer
+ * @private
+ * @param options
+ * @param {BufferType} options.type
+ * @param {Object.<string, BufferAttribute>} options.attributes
+ */
+function Buffer(options) {
+
+    // Create a new Buffer
+    if (!options.isSerializedMapboxBuffer) {
+
+        this.type = options.type;
+        this.size = align(Buffer.SIZE_DEFAULT, Buffer.SIZE_ALIGNMENT);
+        this.length = 0;
+        this.arrayBuffer = new ArrayBuffer(this.size);
+        this.attributes = {};
+        this.itemSize = 0;
+
+        this._refreshArrayBufferViews();
+
+        // Vertex buffer attributes must be aligned to "word" boundaries (4 bytes) but element
+        // buffer attributes do not need to be aligned.
+        var attributeAlignment;
+        if (this.type === Buffer.BufferType.VERTEX) {
+            attributeAlignment = Buffer.VERTEX_ATTRIBUTE_OFFSET_ALIGNMENT;
+        } else {
+            attributeAlignment = null;
+        }
+
+        // Normalize the attributes
+        for (var key in options.attributes) {
+            var attribute = util.extend({}, options.attributes[key]);
+
+            attribute.name = attribute.name || key;
+            attribute.components = attribute.components || 1;
+            attribute.type = attribute.type || Buffer.AttributeType.UNSIGNED_BYTE;
+            attribute.size = attribute.type.size * attribute.components;
+            attribute.offset = this.itemSize;
+
+            this.itemSize = align(attribute.offset + attribute.size, attributeAlignment);
+
+            this.attributes[attribute.name] = attribute;
+        }
+
+    // Restore a serialized buffer
+    } else {
+        var clone = options;
+
+        this.type = clone.type;
+        this.size = clone.size;
+        this.length = clone.index;
+        this.arrayBuffer = clone.arrayBuffer;
+        this.attributes = clone.attributes;
+        this.itemSize = clone.itemSize;
+
+        this._refreshArrayBufferViews();
+    }
+
+    util.assert(this.type);
```

Comment:
This is an overloaded constructor. Like `Buffer1`, it can either

 - create an empty buffer from a configuration object
 - rehydrate a serialized instance of `Buffer`

The idea with the `util.assert(this.type)` line is that we allow each case to parse the constructor arguments independently and then perform a little validation on the resulting instance. It is not necessary to perform this assertion but I found that `this.type` had a tendency to end up as `undefined` and that resulted in hard-to-debug errors.

In a larger sense, I sorta want to rewrite this whole constructor to eliminate this branching altogether. It should be possible to combine these two cases into one.